<launch>
    <!--arg name="mav_name" default="hummingbird"/-->
    <arg name="x" default="0"/>
    <arg name="y" default="0"/>
    <arg name="z" default="0"/>
    <arg name="mav_name" default="firefly"/>
    <arg name="node_type" default="uav"/>
    <arg name="node_id"/>
    <arg name="endpoint"/>
    <arg name="port"/>
    <arg name="min_nodes_num"/>
    <arg name="secret"/>

    <!-- env variables-->
    <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models"/>
    <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models"/>
    <!-- Robot namespace-->
    <group ns="$(arg node_id)">
        <!--Spawn the robot-->
        <include file="$(find rotors_gazebo)/launch/spawn_mav.launch">
            <arg name="mav_name" value="$(arg mav_name)" />
            <arg name="model" value="$(find rotors_description)/urdf/mav_generic_odometry_sensor.gazebo" />
            <arg name="enable_logging" value="false" />
            <arg name="enable_ground_truth" value="true" />
            <arg name="log_file" value="$(arg mav_name)"/>
            <arg name="namespace" value="$(arg node_id)"/>
            <arg name="x" value="$(arg x)"/>
            <arg name="y" value="$(arg y)"/>
            <arg name="z" value="0.0"/>
        </include>
        
        <!-- Load the controller -->
        <node name="lee_position_controller_node" pkg="rotors_control" type="lee_position_controller_node" output="screen">
            <rosparam command="load" file="$(find rotors_gazebo)/resource/lee_controller_$(arg mav_name).yaml" />
            <rosparam command="load" file="$(find rotors_gazebo)/resource/$(arg mav_name).yaml" />
            <remap from="odometry" to="odometry_sensor1/odometry" />
        </node>
        <!-- Start flying -->
        <node name="hovering_example" pkg="rotors_gazebo" type="hovering_example" output="screen">
            <param name="x" value="$(arg x)"/>
            <param name="y" value="$(arg y)"/>
            <param name="z" value="$(arg z)"/>
        </node>

        <node name="navigator" pkg="multirobot_sim" type="navigator.py" output="screen">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="odom_topic" value="/$(arg node_id)/odometry_sensor1/odometry"/>
            <param name="goal_topic" value="goal"/>
            <param name="path_topic" value="path"/>
        </node>
        
        <!-- Repeater node-->
        <node name="repeater" pkg="multirobot_sim" type="repeater.py" output="screen">
            <param name="model_name" value="$(arg node_id)"/>
            <param name="source_topic" value="goal"/>
            <param name="target_topic" value="command/pose"/>
            <param name="path_topic" value="path"/>
        </node>
        <!-- Task allocation-->
        <node name="task_allocation" pkg="multirobot_sim" type="task_allocation.py" output="screen">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
            <param name="odom_topic" value="/$(arg node_id)/odometry_sensor1/odometry"/>
        </node>
        <!-- robot state publisher-->
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
        <!-- joint state publisher-->
        <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />
    </group>
    <include file="$(find multirobot_sim)/launch/blockchain.launch">
            <arg name="node_id" value="$(arg node_id)"/>
            <arg name="node_type" value="$(arg node_type)"/>
            <arg name="endpoint" value="$(arg endpoint)"/>
            <arg name="port" value="$(arg port)"/>
            <arg name="min_nodes_num" value="$(arg min_nodes_num)"/>
            <arg name="secret" value="$(arg secret)"/>
        </include>
</launch>
